questionnaire_linking.py

```diff
@@ -0,0 +1,117 @@
+import csv
+import argparse
+import requests
+import qid_checksum_validator
+from config import Config
+from datetime import datetime
+import uuid
+import json
+from utilities.rabbit_context import RabbitContext
+
+
+def validate_qid_link_file(qid_link_file_path):
+    try:
+        with open(qid_link_file_path, encoding="utf-8") as qid_link_file:
+            return qid_file(qid_link_file)
+    except UnicodeDecodeError as err:
+        print(f'Invalid file encoding, requires utf-8, error: {err}')
+        exit(-1)
+
+
+def qid_file(qid_link_file):
+    qid_link_file_reader = csv.DictReader(qid_link_file, delimiter=',', fieldnames=['case_ref', 'qid'])
+    validate_and_submit_questionnaire_links(qid_link_file_reader)
+
+
+def parse_arguments():
+    parser = argparse.ArgumentParser(description='Load a QID Link file into response management.')
+    parser.add_argument('qid_link_file_path', help='path to the QID Link file', type=str)
+    return parser.parse_args()
+
+
+args = parse_arguments()
+error_count = 0
+
+
+def validate_and_submit_questionnaire_links(qid_link_file_reader):
+    global error_count
+    for line_number, questionnaire_link in enumerate(qid_link_file_reader, 1):
+        error_count = 0
+        for i in questionnaire_link:
+            questionnaire_link[i] = questionnaire_link[i].replace(" ", "")
+        validate_qid_len_and_type(line_number, questionnaire_link["qid"])
+        validate_qid_third_digit(line_number, questionnaire_link["qid"])
+        validate_check_digits(line_number, questionnaire_link["qid"])
+        case_id = validate_case_ref(line_number, questionnaire_link["case_ref"])
+
+        if error_count > 0:
+            print(f'Error: {questionnaire_link} on line {line_number} have FAILED')
+        else:
+            post_message_to_queue(case_id, questionnaire_link["qid"], line_number)
+        print("\n")
+
+
+def validate_case_ref(line_number, case_ref):
+    global error_count
+    if len(case_ref) != 8 or not case_ref.isdigit():
+        print(f'Error: Either wrong length or type located on Case Reference {case_ref}, line {line_number}')
+        error_count += 1
+    if error_count == 0:
```

Comment:
this means that no caserefs will be checked to see if they exist after an error. Is that the desired behaviour, or did we want it to process whatever it could and output ALL the problems?

Reply:
If the case ref isn't the right length or contains non-numerics then we already know it's not going to exist so there's no need to make the API call.

Reply:
I suggest that you put the rest of this function in an `else` block, so that it doesn't bomb out due to one single bad caseref

Reply:
Just to ensure we are on the same page, if there is an error (for example with the QID) in a row of the file, it wont check the case reference however it will then loop to the next line in the file and if that has no errors in it it will check the case reference. If there are no errors in a line then it will produce a success message showing the case ID and confirming the QID number.